### PR TITLE
Send page reload message to android runtime

### PR DESCRIPTION
**Pull request by Fatme**

### CODE REVIEW

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: services/usb-livesync-service-base.ts [hunk 1]
```diff
@@ -155,45 +131,61 @@ export class UsbLiveSyncServiceBase implements IUsbLiveSyncServiceBase {
...
+			this.$logger.info("Transferring project files...");
+			device.fileSystem.transferFiles(deviceAppData.appIdentifier, localToDevicePaths).wait();
+			this.$logger.info("Successfully transferred all project files.");
+		}).future<void>()();
+	}
+
+	private syncCore(data: ILiveSyncData, projectFiles: string[]): IFuture<void> {
+		return (() => {
+			let platform = data.platform ? this.$mobileHelper.normalizePlatformName(data.platform) : this.$devicesService.platform;
+			let deviceAppData =  this.$deviceAppDataFactory.create(data.appIdentifier, this.$mobileHelper.normalizePlatformName(data.platform));
```
Comment (teobugslayer) [inline services/usb-livesync-service-base.ts]: On the previous line, platform was set to `this.$mobileHelper.normalizePlatformName(data.platform)` and therefore can be used here for shorter code.

File: services/usb-livesync-service-base.ts [hunk 2]
```diff
@@ -155,45 +131,61 @@ export class UsbLiveSyncServiceBase implements IUsbLiveSyncServiceBase {
...
+			let localToDevicePaths = this.createLocalToDevicePaths(data.platform, data.appIdentifier, data.localProjectRootPath || data.projectFilesPath, projectFiles);
```
Comment (teobugslayer) [inline services/usb-livesync-service-base.ts]: you should use `platform` (the local variable) instead of `data.platform`, which may be null or undefined.

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: services/usb-livesync-service-base.ts [hunk 3]
```diff
@@ -155,45 +131,62 @@ export class UsbLiveSyncServiceBase implements IUsbLiveSyncServiceBase {
...
+			device.fileSystem.transferFiles(deviceAppData.appIdentifier, localToDevicePaths, projectFilesPath).wait();
+			this.$logger.info("Successfully transferred all project files.");
+		}).future<void>()();
+	}
+
+	private syncCore(data: ILiveSyncData, projectFiles: string[]): IFuture<void> {
+		return (() => {
+			let projectFilesPath = data.localProjectRootPath || data.projectFilesPath;
+			let platform = data.platform ? this.$mobileHelper.normalizePlatformName(data.platform) : this.$devicesService.platform;
+			let deviceAppData =  this.$deviceAppDataFactory.create(data.appIdentifier, this.$mobileHelper.normalizePlatformName(data.platform));
```
Comment (teobugslayer) [inline services/usb-livesync-service-base.ts]: Should we use `platform` here?

File: services/usb-livesync-service-base.ts [hunk 4]
```diff
@@ -155,45 +131,62 @@ export class UsbLiveSyncServiceBase implements IUsbLiveSyncServiceBase {
...
+			let localToDevicePaths = this.createLocalToDevicePaths(data.platform, data.appIdentifier, projectFilesPath, projectFiles);
```
Comment (teobugslayer) [inline services/usb-livesync-service-base.ts]: Should we use `platform` here?

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:broken_heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: mobile/ios/ios-emulator-services.ts [hunk 5]
```diff
@@ -171,5 +172,21 @@ class IosEmulatorServices implements Mobile.IiOSSimulatorService {
...
+
+	private getApplicationPath(appIdentifier: string): string {
+		let iosSim = require("ios-sim-portable");
+		let runningSimulatorId = this.getRunningSimulatorId(appIdentifier);
+		let applicationPath = iosSim.getApplicationPath(runningSimulatorId, appIdentifier);
+		return applicationPath;
+	}
+
+	private getRunningSimulatorId(appIdentifier: string): string {
+		let iosSim = require("ios-sim-portable");
```
Comment (rosen-vladimirov) [inline mobile/ios/ios-emulator-services.ts]: can we move this to property with getter?

File: services/usb-livesync-service-base.ts [hunk 6]
```diff
@@ -78,70 +78,48 @@ export class UsbLiveSyncServiceBase implements IUsbLiveSyncServiceBase {
...
+			let synciOSSimulator = this.shouldSynciOSSimulator(data.platform).wait();
 
 			if(synciOSSimulator) {
-				this.$iOSEmulatorServices.sync(appIdentifier, projectFilesPath, notRunningiOSSimulatorAction).wait();
+				this.$iOSEmulatorServices.sync(data.appIdentifier, data.projectFilesPath, data.notRunningiOSSimulatorAction).wait();
 			}
 
-			if(!this._initialized && (!this.$options.emulator || platform.toLowerCase() === "android")) {
-				this.initialize(platform).wait();
+			if(!this._initialized && (!this.$options.emulator || data.platform.toLowerCase() === "android")) {
```
Comment (rosen-vladimirov) [inline services/usb-livesync-service-base.ts]: you can move this if inside the one on line 93, so the code will be:
```TypeScript
if(!this.$options.emulator || data.platform.toLowerCase() === "android") {
    if(!this._initialized) {
        this.initialize(data.platform).wait();
    }
    let projectFiles = this.$fs.enumerateFilesInDirectorySync(data.projectFilesPath,....

```

File: services/usb-livesync-service-base.ts [hunk 7]
```diff
@@ -78,70 +78,48 @@ export class UsbLiveSyncServiceBase implements IUsbLiveSyncServiceBase {
...
-										iOSSimulatorRelativeToProjectBasePathAction
-									);
+							if(!that.isFileExcluded(filePath, data.excludedProjectDirsAndFiles, data.projectFilesPath)) {
+								let canExecuteFastLiveSync = data.canExecuteFastLiveSync && data.canExecuteFastLiveSync(filePath);
+
+								if(synciOSSimulator && !canExecuteFastLiveSync) {
+									that.batchSimulatorLiveSync(data, filePath);
+								}
+
+								if( (!that.$options.emulator || data.platform.toLowerCase() === "android") && !canExecuteFastLiveSync) {
```
Comment (rosen-vladimirov) [inline services/usb-livesync-service-base.ts]: `(!that.$options.emulator || data.platform.toLowerCase() === "android")` is used many times, can we extract it to a variable

Comment (rosen-vladimirov) [inline services/usb-livesync-service-base.ts]: in fact AppBuilder CLI never uses `this.$options.emulator`, can we pass the value as parameter and remove the `emulator` from common options?

File: mobile/ios/ios-emulator-services.ts [hunk 8]
```diff
@@ -57,20 +57,27 @@ class IosEmulatorServices implements Mobile.IiOSSimulatorService {
 		return this.$childProcess.exec(`${nodeCommandName} ${iosSimPath} ${opts.join(' ')}`);
 	}
 
-	public sync(appIdentifier: string, projectFilesPath: string, notRunningSimulatorAction: () => IFuture<boolean>): IFuture<void> {
-		let syncAction = (applicationPath: string) => shell.cp("-Rf", projectFilesPath, applicationPath);
+	public sync(appIdentifier: string, projectFilesPath: string, notRunningSimulatorAction: () => IFuture<void>): IFuture<void> {
```
Comment (teobugslayer) [inline mobile/ios/ios-emulator-services.ts]: Can we make this function protected?

File: services/usb-livesync-service-base.ts [hunk 9]
```diff
@@ -39,7 +39,7 @@ class SyncBatch {
 					});
 				}
 				this.timer = null;
-			}, 500);
+			}, 250);
```
Comment (rosen-vladimirov) [inline services/usb-livesync-service-base.ts]: why 250, not 125 ? comment or constant could be helpful for the next time we check this code

Comment (ligaz) [inline services/usb-livesync-service-base.ts]: Someone else has already figured it out: https://github.com/Microsoft/TypeScript/blob/master/src/compiler/tsc.ts#L487-L489

Comment (rosen-vladimirov) [inline services/usb-livesync-service-base.ts]: And there's a comment there, I'm sure I'll wonder why we've chosen 250 next time I check this code. Also the explanation there doesn't give exact hints why 250 is the magical number, but I'll be happy in case we have the same comment :)

Comment (ligaz) [inline services/usb-livesync-service-base.ts]: 👣

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

File: services/usb-livesync-service-base.ts (hunk 7 above)
Comment (Fatme) [inline services/usb-livesync-service-base.ts]: It's a good idea but I think this should be done in another PR

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by justcodebuilduser** [comment]
:heart:

**Comment by rosen-vladimirov** [comment]
:+1:  after green build

**Comment by justcodebuilduser** [comment]
:heart:
